### PR TITLE
Change our build process to upload build products for further testing.

### PR DESCRIPTION
**Pull request by ezyang**
See travis/README.md for the executive summary of how this works

Signed-off-by: Edward Z. Yang <ezyang@cs.stanford.edu>

### CODE REVIEW

File: travis/id_rsa [hunk 1]
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
```
Comment (23Skidoo) [inline travis/id_rsa]: Um, shouldn't this file be encrypted?

Comment (23Skidoo) [inline travis/id_rsa]: Relevant links:
    * https://developer.github.com/guides/managing-deploy-keys/
    * https://docs.travis-ci.com/user/encrypting-files/

See also how pushing to the `cabal-website` repo is implemented.

Comment (ezyang) [inline travis/id_rsa]: It's purposely not encrypted so that PRs can take advantage of the infrastructure. PRs don't get access to encrypted files.

Comment (23Skidoo) [inline travis/id_rsa]: OK, so this means that we're giving the whole wide world write access to https://github.com/haskell-pushbot/cabal-binaries
Guess there's no way around this in this scheme.

Comment (ezyang) [inline travis/id_rsa]: Yeah, unfortunately not :(

**Comment by ezyang** [comment]
Agh, webhook not working. Investigating.

**Comment by ezyang** [comment]
Aha! sake-bot is not actually installed on the cabal repository yet. Pinging @hvr

File: travis/id_rsa (hunk 1 above)
Comment (23Skidoo) [inline travis/id_rsa]: OK. Just hoping it won't fill up with spam about Viagra and Cialis.

**Comment by ezyang** [comment]
And it's installed! :)

File: travis-script.sh [hunk 2]
```diff
@@ -107,21 +107,6 @@ if [ "x$CABAL_INSTALL_ONLY" != "xYES" ] ; then
       timed cabal new-build $jobs Cabal Cabal:unit-tests
     fi
 
-    # NB: the '|| exit $?' workaround is required on old broken versions of bash
```
Comment (23Skidoo) [inline travis-script.sh]: Do we use this workaround anywhere else? If so, we should preserve the comment.

File: travis-script.sh [hunk 3]
```diff
@@ -200,3 +157,6 @@ unset CABAL_BUILDDIR
 
 # Check what we got
 ${CABAL_INSTALL_BDIR}/build/cabal/cabal --version
+
+# If this fails, we WANT to fail, because the tests will not be running then
+(timed ./travis/upload.sh) || exit $?
```
Comment (23Skidoo) [inline travis-script.sh]: Looks like we do.

File: travis/README.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+This folder contains scripts for running CI on Travis.
+
+The most unusual thing about our Travis setup is that, after we finish
+building our project, we upload the build products to Git (via the
+haskell-pushbot account) to a separate repository to do testing.
+There are two reasons we do this:
+
+1. On our slowest configuration (GHC 8 on Mac OS X), the time to
+   build and run tests was easily bumping up against the Travis time
+   limit.  By uploading our build products to a separate account
```
Comment (23Skidoo) [inline travis/README.md]: >  By uploading our build products to a separate account

You accidentally a sentence.

File: travis-script.sh (hunk 2 above)
Comment (ezyang) [inline travis-script.sh]: We do, I'll add it back.

**Pull request merged**